dig_test.go

```diff
@@ -0,0 +1,360 @@
+		}), "expected relying on concrete type to succeed")
+
+	})
+
+	t.Run("interface constructor", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(func() io.Writer {
+			return &bytes.Buffer{}
+		}), "provide failed")
+		require.NoError(t, c.Invoke(func(w io.Writer) {
+			require.NotNil(t, w, "invoke got nil interface")
+		}), "invoke failed")
+	})
+
+	t.Run("multiple-type constructor", func(t *testing.T) {
+		c := New()
+		constructor := func() (*bytes.Buffer, []int, error) {
+			return &bytes.Buffer{}, []int{42}, nil
+		}
+		consumer := func(b *bytes.Buffer, nums []int) {
+			assert.NotNil(t, b, "invoke got nil buffer")
+			assert.Equal(t, 1, len(nums), "invoke got empty slice")
+		}
+		require.NoError(t, c.Provide(constructor), "provide failed")
+		require.NoError(t, c.Invoke(consumer), "invoke failed")
+	})
+
+	t.Run("collections and instances of same type", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(func() []*bytes.Buffer {
+			return []*bytes.Buffer{{}}
+		}), "providing collection failed")
+		require.NoError(t, c.Provide(func() *bytes.Buffer {
+			return &bytes.Buffer{}
+		}), "providing pointer failed")
+	})
+}
+
+func TestProvideRespectsConstructorErrors(t *testing.T) {
+	t.Run("constructor succeeds", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(func() (*bytes.Buffer, error) {
+			return &bytes.Buffer{}, nil
+		}), "provide failed")
+		require.NoError(t, c.Invoke(func(b *bytes.Buffer) {
+			require.NotNil(t, b, "invoke got nil buffer")
+		}), "invoke failed")
+	})
+	t.Run("constructor fails", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(func() (*bytes.Buffer, error) {
+			return nil, errors.New("oh no")
+		}), "provide failed")
+
+		var called bool
+		err := c.Invoke(func(b *bytes.Buffer) { called = true })
+		assert.Contains(t, err.Error(), "oh no", "expected to bubble up constructor error")
+		assert.False(t, called, "shouldn't call invoked function when deps aren't available")
+	})
+}
+
+func TestCantProvideUntypedNil(t *testing.T) {
+	t.Fatalf("Calling Container.Provide(nil) crashes the main testing thread. Failing early.")
+	t.Parallel()
+	c := New()
+	assert.Error(t, c.Provide(nil))
+}
+
+func TestCantProvideErrors(t *testing.T) {
+	t.Parallel()
+	c := New()
+
+	assert.Error(t, c.Provide(func() error { return errors.New("foo") }))
+	// TODO: This is another case where we're losing type information, which
+	// (again) makes the provide-instance path behave differently from the
+	// provide-constructor path. This is fixable by not allowing users to
+	// provide types that *implement* error, but let's discuss a holistic
+	// solution.
+	assert.NoError(t, c.Provide(errors.New("foo")))
+}
+
+func TestProvideKnownTypesFails(t *testing.T) {
+	t.Parallel()
+	c := New()
+	assert.NoError(t, c.Provide(func() *bytes.Buffer { return nil }))
+	assert.Error(t, c.Provide(func() *bytes.Buffer { return nil }))
```

Comment:
can you verify that it fails even if the value is provided. this might need to be a loop test like:
```
provideArgs := []interface{}{
  func() *bytes.Buffer { return nil },
  func() (*bytes.Buffer, error) { return nil, nil },
  &bytes.Buffer{},
}

for _, firstProvide := range provideArgs {
  assert.NoError(t, c.Provide(firstProvide), ...)
  for _, secondProvide := range provideArgs {
    assert.Error(t, c.Provide(secondProvide), ...)
  }
}
```

that way we ensure that once a type is known (regardless of how it was provided), other provides fail.